peer.go

```diff
@@ -938,137 +984,182 @@ func (p *peer) handleLocalClose(req *htlcswitch.ChanClose) {
-	feeRate := p.server.cc.feeEstimator.EstimateFeePerWeight(1) * 1000
+	feeRate := btcutil.Amount(
+		p.server.cc.feeEstimator.EstimateFeePerWeight(1) * 1000)
 
-	// TODO(roasbeef): actually perform fee negotiation here, only send sig
-	// if we agree to fee
-
-	// Once both sides agree on a fee, we'll create a signature that closes
-	// the channel using the agree upon fee rate.
+	// We propose a fee and send a close proposal to the peer. This will
+	// start the fee negotiations. Once both sides agree on a fee, we'll
+	// create a signature that closes the channel using the agreed upon fee.
 	// TODO(roasbeef): remove encoding redundancy
-	closeSig, proposedFee, err := channel.CreateCloseProposal(feeRate)
+	fee := channel.CalcFee(feeRate)
+	closeSig, proposedFee, err := channel.CreateCloseProposal(fee)
 	if err != nil {
 		peerLog.Errorf("unable to create close proposal: %v", err)
-		return nil
+		return nil, 0
 	}
 	parsedSig, err := btcec.ParseSignature(closeSig, btcec.S256())
 	if err != nil {
 		peerLog.Errorf("unable to parse signature: %v", err)
-		return nil
+		return nil, 0
 	}
 
 	// With the closing signature assembled, we'll send the matching close
 	// signed message to the other party so they can broadcast the closing
-	// transaction.
+	// transaction if they agree with the fee, or create a new close
+	// proposal if they don't.
 	closingSigned := lnwire.NewClosingSigned(msg.ChannelID, proposedFee,
 		parsedSig)
 	p.queueMsg(closingSigned, nil)
 
-	return closeSig
+	return closeSig, proposedFee
+}
+
+// calculateCompromiseFee performs the current fee negotiation algorithm,
+// taking into consideration our ideal fee based on current fee environment,
+// the fee we last proposed (if any), and the fee proposed by the peer.
+func calculateCompromiseFee(ourIdealFee, lastSentFee, peerFee uint64) uint64 {
+	// We will accept a proposed fee in the interval
+	// [0.5*ourIdealFee, 2*ourIdealFee]. If the peer's fee doesn't fall in
+	// this range, we'll propose the average of the peer's fee and our last
+	// sent fee, as long as it is in this range.
+	// TODO: Dynamic fee to determine what we consider min/max for timely
+	// confirmation.
+	maxFee := 2 * ourIdealFee
+	minFee := ourIdealFee / 2
+
+	// If we didn't propose a fee before, just use our ideal fee value for
+	// the average calculation.
+	if lastSentFee == 0 {
+		lastSentFee = ourIdealFee
+	}
+	avgFee := (lastSentFee + peerFee) / 2
+
+	switch {
+	case peerFee <= maxFee && peerFee >= minFee:
+		// Peer fee is in the accepted range.
+		return peerFee
+	case avgFee <= maxFee && avgFee >= minFee:
+		// The peer's fee is not in the accepted range, but the average
+		// fee is.
+		return avgFee
+	case avgFee > maxFee:
+		// We cannot accept the average fee, as it is more than twice
+		// our own estimate. Set our proposed to the maximum we can
+		// accept.
+		return maxFee
+	default:
+		// Cannot accept the average, as we consider it too low.
```

Comment:
Added a TODO, but this check should be fairly straightforward to just add in this PR I think. Where do I fetch the current fee? Create a getter for `channel.channelState.CommitFee`?